k2/python/tests/autograd_test.py

```diff
@@ -0,0 +1,294 @@
+            0 2 2 100
+            1 3 -1 3.5
+            2 3 -1 5.5
+            3
+        '''
+
+        fsa1 = k2.Fsa.from_str(s1)
+        fsa2 = k2.Fsa.from_str(s2)
+        fsa3 = k2.Fsa.from_str(s3)
+
+        fsa1.scores.requires_grad_(True)
+        fsa2.scores.requires_grad_(True)
+        fsa3.scores.requires_grad_(True)
+
+        fsa_vec = k2.create_fsa_vec([fsa1, fsa2, fsa3])
+
+        log_like = k2.get_forward_log_like(fsa_vec,
+                                           log_semiring=False,
+                                           use_float_scores=True)
+        assert log_like.dtype == torch.float32
+        expected_log_like = torch.tensor([14, 13, 105.5])
+        assert torch.allclose(log_like, expected_log_like)
+
+        log_like.sum().backward()
+
+        fsa1_best_arc_indexes = torch.tensor([1, 3, 5, 10])
+        assert torch.allclose(fsa1.scores.grad[fsa1_best_arc_indexes],
+                              torch.ones(4, dtype=torch.float32))
+        assert fsa1.scores.grad.sum() == 4
+
+        fsa2_best_arc_indexes = torch.tensor([1, 4, 5, 7])
+        assert torch.allclose(fsa2.scores.grad[fsa2_best_arc_indexes],
+                              torch.ones(4, dtype=torch.float32))
+        assert fsa2.scores.grad.sum() == 4
+
+        fsa3_best_arc_indexes = torch.tensor([1, 3])
+        assert torch.allclose(fsa3.scores.grad[fsa3_best_arc_indexes],
+                              torch.ones(2, dtype=torch.float32))
+        assert fsa3.scores.grad.sum() == 2
+
+        # now for double
+        fsa1.scores.grad = None
+        fsa2.scores.grad = None
+        fsa3.scores.grad = None
+        log_like = k2.get_forward_log_like(fsa_vec,
+                                           log_semiring=False,
+                                           use_float_scores=False)
+
+        assert log_like.dtype == torch.float64
+        expected_log_like = expected_log_like.to(torch.float64)
+        assert torch.allclose(log_like, expected_log_like)
+
+        log_like.sum().backward()
+
+        assert torch.allclose(fsa1.scores.grad[fsa1_best_arc_indexes],
+                              torch.ones(4, dtype=torch.float32))
+        assert fsa1.scores.grad.sum() == 4
+
+        assert torch.allclose(fsa2.scores.grad[fsa2_best_arc_indexes],
+                              torch.ones(4, dtype=torch.float32))
+        assert fsa2.scores.grad.sum() == 4
+
+        assert torch.allclose(fsa3.scores.grad[fsa3_best_arc_indexes],
+                              torch.ones(2, dtype=torch.float32))
+        assert fsa3.scores.grad.sum() == 2
+
+    def test_log_single_fsa(self):
+        s = '''
+            0 1 1 0.1
+            0 2 2 0.2
+            1 2 3 0.3
+            1 3 4 0.4
+            2 3 5 0.5
+            3 4 -1 0
+            4
+        '''
+        fsa = k2.Fsa.from_str(s)
+        fsa.scores.requires_grad_(True)
+        fsa_vec = k2.create_fsa_vec([fsa])
+        log_like = k2.get_forward_log_like(fsa_vec,
```

Comment:
It's great that you have done the autograd tests!
I'm not sure that I like the name `get_forward_log_like`.  I think I'd prefer `get_tot_scores` which is more consistent with the C++ code.  I particularly don't like the 'forward' part of the name, because the log_like is just a property of the FSA (+semiring), and we could just as easily have chosen to implement it via backward.
In fact, it might make sense to do this (implement via backward_loglikes), because that code is likely faster than the forward code.  [We'd have to add a tot_scores arg to the forward-pass to make this work, and also a function to get the tot-scores from the start states, so for now leave it as-is.]

Reply:
Renamed to `get_tot_scores`.